# HW 1-10: Methods of Integration

This task evaluates the accuracy of three numerical methods—Euler’s method, the Midpoint method, and the Runge-Kutta method—for solving the differential equation  $\frac{dx}{dt} = x$  with the initial condition  $x(0) = 1$  over the interval  $t \in [0, 1]$. 

The analytical solution to this equation is  $x(t) = e^t$ , and at  $t = 1$ , the exact solution is  $e$ .

The implementation does the following:

1. **Define the ODE**: The derivative function $ \frac{dx}{dt} = x $.
2. **Define Numerical Methods**:
   - **Euler's Method**: A basic explicit first-order method.
   - **Midpoint Method**: A second-order Runge-Kutta method.
   - **Runge-Kutta Method**: A fourth-order method.
3. **Discretization**: Use step sizes corresponding to $ N = 2^{10}$ to $ N = 2^{20}$, where $N$ represents the number of discretization points.
4. **Error Analysis**: Compare numerical results with the exact solution to compute:
   - Absolute errors.
   - Percentage errors.

In [1]:
import numpy as np
import pandas as pd

# Numerical Integration Techniques
def euler_step(deriv_func, initial_val, time_points, step_size):
    current = initial_val
    for current_time in time_points[:-1]:
        current += step_size * deriv_func(current_time, current)
    return current

def midpoint_step(deriv_func, initial_val, time_points, step_size):
    current = initial_val
    for current_time in time_points[:-1]:
        temp = current + (step_size / 2) * deriv_func(current_time, current)
        current += step_size * deriv_func(current_time + step_size / 2, temp)
    return current

def runge_kutta_step(deriv_func, initial_val, time_points, step_size):
    current = initial_val
    for current_time in time_points[:-1]:
        k1 = deriv_func(current_time, current)
        k2 = deriv_func(current_time + step_size / 2, current + (step_size * k1) / 2)
        k3 = deriv_func(current_time + step_size / 2, current + (step_size * k2) / 2)
        k4 = deriv_func(current_time + step_size, current + step_size * k3)
        current += (step_size / 6) * (k1 + 2*k2 + 2*k3 + k4)
    return current

# Differential equation dx/dt = x
def derivative(t, x):
    return x

# Initial condition
initial_x = 1

# Analytical solution at t=1
exact_solution = np.exp(1)

# Discretization points: from 2^10 to 2^20
num_steps_list = [2**k for k in range(10, 21)]
simulation_results = []

# Perform simulations for each N
for num_steps in num_steps_list:
    delta_t = 1.0 / num_steps
    times = np.linspace(0, 1, num_steps + 1)
    
    # Compute numerical solutions
    euler_val = euler_step(derivative, initial_x, times, delta_t)
    midpoint_val = midpoint_step(derivative, initial_x, times, delta_t)
    rk_val = runge_kutta_step(derivative, initial_x, times, delta_t)
    
    # Calculate absolute errors
    error_euler = abs(euler_val - exact_solution)
    error_midpoint = abs(midpoint_val - exact_solution)
    error_rk = abs(rk_val - exact_solution)
    
    # Calculate percentage errors
    percent_error_euler = (error_euler / exact_solution) * 100
    percent_error_midpoint = (error_midpoint / exact_solution) * 100
    percent_error_rk = (error_rk / exact_solution) * 100
    
    # Append results
    simulation_results.append([
        error_euler,
        error_midpoint,
        error_rk,
        percent_error_euler,
        percent_error_midpoint,
        percent_error_rk
    ])

# Organize results into a DataFrame
result_columns = [
    "Euler Absolute Error",
    "Midpoint Absolute Error",
    "Runge-Kutta Absolute Error",
    "Euler % Error",
    "Midpoint % Error",
    "Runge-Kutta % Error"
]
results_df = pd.DataFrame(simulation_results, index=num_steps_list, columns=result_columns)

# Rename index to "N"
results_df.index.name = "Steps (N)"

# Display the results
display(results_df)

Euler Absolute Error  Midpoint Absolute Error  \
Steps (N)                                                  
1024                   0.001326             4.317429e-07   
2048                   0.000663             1.079753e-07   
4096                   0.000332             2.699876e-08   
8192                   0.000166             6.750315e-09   
16384                  0.000083             1.687626e-09   
32768                  0.000041             4.219323e-10   
65536                  0.000021             1.054659e-10   
131072                 0.000010             2.643086e-11   
262144                 0.000005             6.560086e-12   
524288                 0.000003             1.751044e-12   
1048576                0.000001             5.662137e-13   

           Runge-Kutta Absolute Error  Euler % Error  Midpoint % Error  \
Steps (N)                                                                
1024                     1.953993e-14       0.048784      1.588293e-05   
2048                     3.552714e-15       0.024403      3.972188e-06   
4096                     8.437695e-15       0.012204      9.932289e-07   
8192                     2.131628e-14       0.006103      2.483302e-07   
16384                    2.220446e-15       0.003052      6.208431e-08   
32768                    1.509903e-14       0.001526      1.552202e-08   
65536                    5.462297e-14       0.000763      3.879872e-09   
131072                   5.595524e-14       0.000381      9.723369e-10   
262144                   2.886580e-14       0.000191      2.413321e-10   
524288                   1.332268e-13       0.000095      6.441730e-11   
1048576                  1.167955e-13       0.000048      2.082984e-11   

           Runge-Kutta % Error  
Steps (N)                       
1024              7.188337e-13  
2048              1.306970e-13  
4096              3.104055e-13  
8192              7.841822e-13  
16384             8.168565e-14  
32768             5.554624e-13  
65536             2.009467e-12  
131072            2.058478e-12  
262144            1.061913e-12  
524288            4.901139e-12  
1048576           4.296665e-12

#### Key Observations:

1. **Accuracy Improves with $N$**:
   - Across all methods, the absolute and percentage errors decrease as $ N $ increases. This demonstrates the expected behavior that reducing the step size improves numerical accuracy.

2. **Runge-Kutta Outperforms Other Methods**:
   - Runge-Kutta consistently shows significantly lower errors compared to Euler and Midpoint methods. This is expected as Runge-Kutta is a fourth-order method, providing greater accuracy for the same step size.

3. **Midpoint vs. Euler**:
   - Midpoint, being a second-order method, performs better than Euler's method, which is first-order. For instance, at $ N = 1024 $, Midpoint's absolute error is several orders of magnitude smaller than Euler's.

4. **Diminishing Returns at Large $ N $**:
   - The rate of error reduction slows down for larger $N$. For example, moving from $ N = 524288$ to $ N = 1048576 $, the improvement in error becomes minimal, particularly for Runge-Kutta. This indicates that beyond a certain point, reducing the step size yields diminishing returns due to limitations such as numerical precision.

#### Conclusion

- **Runge-Kutta is the most accurate method** and should be preferred when precision is critical.
- **Midpoint strikes a balance between accuracy and computational cost**, making it suitable for moderately accurate requirements.
- **Euler's method is the least accurate**, but its simplicity makes it a good choice for basic approximations or when computational resources are limited.